pkg/kube/build_lock.go

```diff
@@ -0,0 +1,415 @@
+package kube
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/jenkins-x/jx/pkg/log"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+)
+
+// Labels required to be a lock. Anything else should be ignored.
+var buildLockLabels map[string]string = map[string]string{
+	"jenkins-x.io/kind": "build-lock",
+}
+
+// AcquireBuildLock acquires a build lock, to avoid other builds to edit the
+// same namespace while a deployment is already running, other deployments
+// can negotiate which one should run after, by editing its data.
+// Returns a function to release the lock (to be called in a defer)
+// Returns an error if a newer build is already running, or if an error happened
+func AcquireBuildLock(kubeClient kubernetes.Interface, devNamespace, namespace string) (func() error, error) {
+	// Only lock if running in Tekton
+	if ok, err := IsTektonEnabled(kubeClient, devNamespace); err != nil {
+		log.Logger().Warnf("error while looking for Tekton: %s\n", err.Error())
+		return nil, err
+	} else if !ok {
+		log.Logger().Infof("lock cancelled because not running in tekton")
+		return func() error { return nil }, nil
+	}
+	// Get infos from the headers
+	now := time.Now().UTC().Format(time.RFC3339Nano)
+	owner := os.Getenv("REPO_OWNER")
+	if owner == "" {
+		log.Logger().Warnf("no REPO_OWNER provided")
+		return nil, fmt.Errorf("no REPO_OWNER provided")
+	}
+	repository := os.Getenv("REPO_NAME")
+	if repository == "" {
+		log.Logger().Warnf("no REPO_NAME provided")
+		return nil, fmt.Errorf("no REPO_NAME provided")
+	}
+	branch := os.Getenv("BRANCH_NAME")
+	if branch == "" {
+		log.Logger().Warnf("no BRANCH_NAME provided")
+		return nil, fmt.Errorf("no BRANCH_NAME provided")
+	}
+	build := os.Getenv("BUILD_NUMBER")
+	if _, err := strconv.Atoi(build); err != nil {
+		log.Logger().Warnf("no BUILD_NUMBER provided: %s\n", err.Error())
+		return nil, err
+	}
+	// Find our pod
+	podList, err := kubeClient.CoreV1().Pods(devNamespace).List(metav1.ListOptions{
```

Comment:
I am struggling a bit why we need to tie the locking via a `ConfigMap` to a running pod? Even though uncommon, I think about the case where the step is executed locally. `jx boot` comes to mind.

Could we just not work with the ConfigMap? I think that's also what @ccojocar suggested in a comment on the issue.

Reply:
The idea to use a ConfigMap is because we need an atomic operation at the Kubernetes level. Only create-update-delete are atomic as far as I know. Looking at the present pods is not a strong guarantee. Plus the ConfigMap is a playground for other job to negotiate which one will be next.

The ConfigMap is linked to a pod such that we can know if the pipeline is still running, in case of the pipeline did not have the opportunity to clean the ConfigMap (for instance with `jx stop pipeline`). If not, the lock could lock every future build.

I agree that this mechanism is not adapted to local commands like `jx boot`. But first those are manual commands, there is no reason for it to be launched 2 times. Second, I think blocking `jx boot` because another one may be running (and we don't have any way to be sure) could be horribly wrong when somebody is trying to fix a failed `jx boot`.

Reply:
> The idea to use a ConfigMap is because we need an atomic operation at the Kubernetes level. Only create-update-delete are atomic as far as I know. 

I am onboard with the ConfigMap. I think that's the right thing to do.

> The ConfigMap is linked to a pod such that we can know if the pipeline is still running,

That's the part I am not so sure about. The linking to a pod. 

> If not, the lock could lock every future build.

Well, right. There is a potential for a lock not getting cleaned up if something crashes, but I was more thinking about working with timeouts. If there is a lock which does not get cleaned up and another build exceeds its timeout it will fail with an appropriate error message. And yes, one needs to then either manually delete the lock (which might be the first approach) or then eventually build something into `jx`. I think that would make things also a bit simpler and solve also the running locally case.

> Second, I think blocking jx boot because another one may be running ...

I am not so sure. There is actually an open issue for this as well. Given that the outcome of two concurrent boot runs would be unclear at best, I'd argue a `jx boot` needs to be locked as well. If you try to "fix" a failed install and there are locking issues, I would expect that you sort them out first. 

@jenkins-x/core thoughts?